README.md
# cluster-client
Sharing Connection among Multi-Process Nodejs

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/cluster-client.svg?style=flat-square
[npm-url]: https://npmjs.org/package/cluster-client
[travis-image]: https://img.shields.io/travis/node-modules/cluster-client.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/cluster-client
[codecov-image]: https://codecov.io/gh/node-modules/cluster-client/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/node-modules/cluster-client
[david-image]: https://img.shields.io/david/node-modules/cluster-client.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/cluster-client
[snyk-image]: https://snyk.io/test/npm/cluster-client/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/cluster-client
[download-image]: https://img.shields.io/npm/dm/cluster-client.svg?style=flat-square
[download-url]: https://npmjs.org/package/cluster-client

As we know, each Node.js process runs in a single thread. Usually, we split a single process into multiple processes to take advantage of multi-core systems. On the other hand, it brings more system overhead, sush as maintaining more TCP connections between servers.

This module is designed to share connections among multi-process Nodejs.

## Theory

- Inspired by [Leader/Follower pattern](http://www.cs.wustl.edu/~schmidt/PDF/lf.pdf).
- Allow ONLY one process "the Leader" to communicate with server. Other processes "the Followers" act as "Proxy" client, and forward all requests to Leader.
- The Leader is selected by "Port Competition". Every process try to listen on a certain port (for example 7777), but ONLY one can occupy the port, then it becomes the Leader, the others become Followers.
- TCP socket connections are maintained between Leader and Followers. And I design a simple communication protocol to exchange data between them.
- If old Leader dies, one of processes will be selected as the new Leader.

## Diagram

normal (without using cluster client)
```js
+--------+   +--------+
| Client |   | Client |   ...
+--------+   +--------+
    |  \     /   |
    |    \ /     |
    |    / \     |
    |  /     \   |
+--------+   +--------+
| Server |   | Server |   ...
+--------+   +--------+

```

using cluster-client
```js
             +-------+
             | start |
             +---+---+
                 |
        +--------+---------+
      __| port competition |__
win /   +------------------+  \ lose
   /                           \
+--------+     tcp conn     +----------+
| Leader |<---------------->| Follower |
+--------+                  +----------+
    |
+--------+
| Client |
+--------+
    |  \
    |    \
    |      \
    |        \
+--------+   +--------+
| Server |   | Server |   ...
+--------+   +--------+

```

## Protocol

- Packet structure
```js
 0       1       2               4                                                              12
 +-------+-------+---------------+---------------------------------------------------------------+
 |version|req/res|    reserved   |                          request id                           |
 +-------------------------------+-------------------------------+-------------------------------+
 |           timeout             |   connection object length    |   application object length   |
 +-------------------------------+---------------------------------------------------------------+
 |         conn object (JSON format)  ...                    |            app object             |
 +-----------------------------------------------------------+                                   |
 |                                          ...                                                  |
 +-----------------------------------------------------------------------------------------------+
```
- Protocol Type
  - Register Channel
  - Subscribe/Publish
  - Invoke
- Sequence diagram

```js
 +----------+             +---------------+          +---------+
 | Follower |             |  local server |          |  Leader |
 +----------+             +---------------+          +---------+
      |     register channel     |       assign to        |
      + -----------------------> |  --------------------> |
      |                          |                        |
      |                                subscribe          |
      + ------------------------------------------------> |
      |       subscribe result                            |
      | <------------------------------------------------ +
      |                                                   |
      |                                 invoke            |
      + ------------------------------------------------> |
      |          invoke result                            |
      | <------------------------------------------------ +
      |                                                   |
```

## Install

```bash
$ npm install cluster-client --save
```

Node.js >= 6.0.0 required

## Usage

```js
'use strict';

const co = require('co');
const Base = require('sdk-base');
const cluster = require('cluster-client');

/**
 * Client Example
 */
class YourClient extends Base {
  constructor(options) {
    super(options);

    this.options = options;
    this.ready(true);
  }

  subscribe(reg, listener) {
    // subscribe logic
  }

  publish(reg) {
    // publish logic
  }

  * getData(id) {
    // invoke api
  }

  getDataCallback(id, cb) {
    // ...
  }

  getDataPromise(id) {
    // ...
  }
}

// create some client instances, but only one instance will connect to server
const client_1 = cluster(YourClient)
  .delegate('getData')
  .delegate('getDataCallback')
  .delegate('getDataPromise')
  .create({ foo: 'bar' });
const client_2 = cluster(YourClient)
  .delegate('getData')
  .delegate('getDataCallback')
  .delegate('getDataPromise')
  .create({ foo: 'bar' });
const client_3 = cluster(YourClient)
  .delegate('getData')
  .delegate('getDataCallback')
  .delegate('getDataPromise')
  .create({ foo: 'bar' });

// subscribe information
client_1.subscribe('some thing', result => console.log(result));
client_2.subscribe('some thing', result => console.log(result));
client_3.subscribe('some thing', result => console.log(result));

// publish data
client_2.publish('some data');

// invoke method
client_3.getDataCallback('some thing', (err, val) => console.log(val));
client_2.getDataPromise('some thing').then(val => console.log(val));

co(function*() {
  const ret = yield client_1.getData('some thing');
  console.log(ret);
}).catch(err => console.error(err));
```

## API

- `delegate(from, to)`:
  create delegate method, `from` is the method name your want to create, and `to` have 3 possible values: [ `subscribe`, `unSubscribe`, `publish`, `invoke`, `invokeOneway`, `close` ],  and the default value is invoke
- `override(name, value)`:
  override one property
- `create(…)`
  create the client instance
- `close(client)`
  close the client


[MIT](LICENSE)

History.md

3.3.1 / 2022-11-23
==================

**fixes**
  * [[`9d9b965`](http://github.com/node-modules/cluster-client/commit/9d9b965a50e148bff78c8763bcd4dea7e2998e57)] - fix: should not emit timeout when has cache (#58) (killa <<killa123@126.com>>)

3.3.0 / 2022-11-14
==================

**features**
  * [[`8ed198d`](http://github.com/node-modules/cluster-client/commit/8ed198db8d64eb76e018699093faed7c88f8c412)] - feat: add options isCheckHeartbeat of whether check heartbeat (#57) (sinkhaha <<1468709106@qq.com>>)

3.2.0 / 2022-10-18
==================

**features**
  * [[`c9f389d`](http://github.com/node-modules/cluster-client/commit/c9f389d277abf75a5b3274626b326b6642b208f1)] - feat: impl subscribeTimeout (#56) (killa <<killa123@126.com>>)

3.1.1 / 2022-06-22
==================

**fixes**
  * [[`f4cc11b`](http://github.com/node-modules/cluster-client/commit/f4cc11bb539f58cc4926f1ce10db6b4ae1e7769c)] - fix: subscribe is not working on single thread mode (#51) (钟典 Desmond <<z@limme.net>>)

3.1.0 / 2021-11-21
==================

**features**
  * [[`a21dde4`](http://github.com/node-modules/cluster-client/commit/a21dde4e7e20b232fe62c44266a0391b4eb4e46a)] - feat: connectTimeout support env, default 10s (#55) (mansonchor.github.com <<mansonchor1987@gmail.com>>)

3.0.1 / 2019-03-01
==================

**fixes**
  * [[`6f9bf8e`](http://github.com/node-modules/cluster-client/commit/6f9bf8e300386f28c2a208aa8c394d441045c03c)] - fix: single mode bugs (#49) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

3.0.0 / 2019-02-26
==================

**features**
  * [[`7f3765d`](http://github.com/node-modules/cluster-client/commit/7f3765ded8877758c9a0bf0c1b23ab4e2c932c21)] - feat: support cpu single mode (#48) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

2.1.2 / 2018-11-23
==================

**others**
  * [[`4c5093b`](http://github.com/node-modules/cluster-client/commit/4c5093b878ee8018524b1b0a514865e647a8f38c)] - chore: upgrade deps (#47) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

2.1.1 / 2018-06-12
==================

**fixes**
  * [[`9fe3849`](http://github.com/node-modules/cluster-client/commit/9fe38494d41afca76491c78cf54b2717508f94b9)] - fix: follower should not retry register channel if socket already disconnected (#43) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

2.1.0 / 2018-03-26
==================

**features**
  * [[`dce3fee`](http://github.com/node-modules/cluster-client/commit/dce3fee89a5c3d617b09e129f3ee68214fa90ccd)] - feat: response with whole error object instead of only stack, message (#38) (killa <<killa123@126.com>>)

2.0.0 / 2018-03-06
==================

**others**
  * [[`2574eae`](http://github.com/node-modules/cluster-client/commit/2574eae6fdbe603b74a3c27a57e3b545aec54314)] - [BREAKING] feat: migrating from generators to async/await (#36) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)
  * [[`e32f0ef`](http://github.com/node-modules/cluster-client/commit/e32f0eff5fe5dfb4187386d928b2fa8cdc65cfa5)] - chore: release 1.7.1 (xiaochen.gaoxc <<xiaochen.gaoxc@alibaba-inc.com>>),

1.7.1 / 2017-09-21
==================

  * fix: error occured while calling invokeOneway method if ready failed (#33)

1.7.0 / 2017-08-18
==================

  * feat: support custom port by env NODE_CLUSTER_CLIENT_PORT (#31)

1.6.8 / 2017-08-18
==================

  * fix: make sure leader ready before follower in egg (#32)

1.6.7 / 2017-07-31
==================

  * fix: only close server when server exists (#30)

1.6.6 / 2017-07-28
==================

  * fix: set exclusive to true on listen (#29)

1.6.5 / 2017-06-24
==================

  * fix: ignore error after close & register channel issue (#28)

1.6.4 / 2017-05-08
==================

  * chore: remove unnecessary log, using debug instead (#27)

1.6.3 / 2017-04-25
==================

  * fix: make sure follower test socket end (#25)

1.6.2 / 2017-04-25
==================

  * fix: ignore ECONNRESET error (#24)

1.6.1 / 2017-04-20
==================

  * fix: invoke before client ready issue (#23)
  * fix: fix symbol property error (#22)

1.6.0 / 2017-04-18
==================

  * feat: make clustClient method writable to support mock or spy (#21)

1.5.4 / 2017-04-12
==================

  * fix: avoid event memory leak warning (#20)

1.5.3 / 2017-03-17
==================

  * fix: make sure subscribe listener triggered asynchronized (#19)

1.5.2 / 2017-03-14
==================

  * fix: event delegate & leader ready bug (#18)

1.5.1 / 2017-03-13
==================

  * fix: don't auto ready when initMethod exists (#17)

1.5.0 / 2017-03-10
==================

  * feat: add APIClientBase to help you create your api client (#16)

1.4.0 / 2017-03-08
==================

  * feat: support unSubscribe, invokeOneway & close self (#14)

1.3.2 / 2017-03-08
==================

  * fix: fix leader subscribe issue & heartbeat timeout issue (#15)

1.3.1 / 2017-03-07
==================

  * chore: better notice (#13)
  * test: fix failed case (#12)

1.3.0 / 2017-02-22
==================

  * fix: block all remote connection (#11)

1.2.0 / 2017-02-20
==================

  * feat: use serialize-json to support encode/decode buffer, date, undef… (#10)

1.1.0 / 2017-02-07
==================

  * feat: close (#7)
  * fix: no more need harmony-reflect on node >= 6 (#8)
  * refactor: improve utils.delegateEvents() (#6)

1.0.3 / 2017-02-04
==================

  * fix: adjust serialize algorithm for invoke arguments (#3)

1.0.2 / 2017-01-25
==================

  * fix: log error if error exist (#5)
  * docs: fix typo subsribe -> subscribe (#4)

1.0.1 / 2016-12-26
==================

  * fix: fix shared memory issue (#2)

1.0.0 / 2016-12-22
==================

  * feat: implement cluster-client


lib/follower.js
'use strict';

const debug = require('debug')('cluster-client#follower');
const is = require('is-type-of');
const Base = require('tcp-base');
const Packet = require('./protocol/packet');
const Request = require('./protocol/request');
const Response = require('./protocol/response');

class Follower extends Base {
  /**
   * "Fake" Client, forward request to leader
   *
   * @param {Object} options
   *  - {Number} port - the port
   *  - {Map} descriptors - interface descriptors
   *  - {Transcode} transcode - serialze / deserialze methods
   *  - {Number} responseTimeout - the timeout
   *  - {Number} subscribeTimeout - the first subscribe callback timeout
   * @class
   */
  constructor(options) {
    // local address
    options.host = '127.0.0.1';
    super(options);
    this._publishMethodName = this._findMethodName('publish');
    this._subInfo = new Set();
    this._subData = new Map();
    this._transcode = options.transcode;
    this._closeByUser = false;
    this._subscribeTimeoutListeners = new Map();

    this.on('request', req => this._handleRequest(req));
    // avoid warning message
    this.setMaxListeners(100);
  }
  get isLeader() {
    return false;
  }

  get logger() {
    return this.options.logger;
  }

  get heartBeatPacket() {
    const heartbeat = new Request({
      connObj: {
        type: 'heartbeat',
      },
      timeout: this.options.responseTimeout,
    });
    return heartbeat.encode();
  }

  getHeader() {
    return this.read(24);
  }

  getBodyLength(header) {
    return header.readInt32BE(16) + header.readInt32BE(20);
  }

  close(err) {
    this._closeByUser = true;
    return super.close(err);
  }

  decode(body, header) {
    const buf = Buffer.concat([ header, body ]);
    const packet = Packet.decode(buf);
    const connObj = packet.connObj;
    if (connObj && connObj.type === 'invoke_result') {
      let data;
      if (packet.data) {
        data = this.options.transcode.decode(packet.data);
      }
      if (connObj.success) {
        return {
          id: packet.id,
          isResponse: packet.isResponse,
          data,
        };
      }
      const error = new Error(data.message);
      Object.assign(error, data);
      return {
        id: packet.id,
        isResponse: packet.isResponse,
        error,
      };
    }
    return {
      id: packet.id,
      isResponse: packet.isResponse,
      connObj: packet.connObj,
      data: packet.data,
    };
  }

  send(...args) {
    // just ignore after close
    if (this._closeByUser) {
      return;
    }
    return super.send(...args);
  }

  formatKey(reg) {
    return '$$inner$$__' + this.options.formatKey(reg);
  }

  _clearSubscribeTimeout(key) {
    if (!this._subscribeTimeoutListeners.has(key)) {
      return;
    }
    const timeout = this._subscribeTimeoutListeners.get(key);
    clearTimeout(timeout);
    this._subscribeTimeoutListeners.delete(key);
  }

  _setSubscribeTimeout(key) {
    this._clearSubscribeTimeout(key);
    const start = Date.now();
    const subscribeTimeout = this.options.subscribeTimeout;
    const timeout = setTimeout(() => {
      const error = new Error(`subscribe timeout for ${key}, cost: ${Date.now() - start}ms, expect is ${subscribeTimeout}ms`);
      this.emit('error', error);
    }, subscribeTimeout);
    this._subscribeTimeoutListeners.set(key, timeout);
  }

  _listenSubscribeTimeout(key) {
    this._setSubscribeTimeout(key);
    const dataListener = () => {
      this.removeListener(key, dataListener);
      this._clearSubscribeTimeout(key);
    };
    this.on(key, dataListener);
  }

  subscribe(reg, listener) {
    const key = this.formatKey(reg);
    this.on(key, listener);
    if (this.options.subscribeTimeout && !this._subData.has(key)) {
      this._listenSubscribeTimeout(key);
    }

    // no need duplicate subscribe
    if (!this._subInfo.has(key)) {
      debug('[Follower:%s] subscribe %j for first time', this.options.name, reg);
      const req = new Request({
        connObj: { type: 'subscribe', key, reg },
        timeout: this.options.responseTimeout,
      });

      // send subscription
      this.send({
        id: req.id,
        oneway: true,
        data: req.encode(),
      });
      this._subInfo.add(key);
    } else if (this._subData.has(key)) {
      debug('[Follower:%s] subscribe %j', this.options.name, reg);
      process.nextTick(() => {
        listener(this._subData.get(key));
      });
    }
    return this;
  }

  unSubscribe(reg, listener) {
    const key = this.formatKey(reg);
    if (listener) {
      this.removeListener(key, listener);
    } else {
      this.removeAllListeners(key);
    }
    if (this.listeners(key).length === 0) {
      debug('[Follower:%s] no more subscriber for %j, send unSubscribe req to leader', this.options.name, reg);
      this._subInfo.delete(key);

      const req = new Request({
        connObj: { type: 'unSubscribe', key, reg },
        timeout: this.options.responseTimeout,
      });
      // send subscription
      this.send({
        id: req.id,
        oneway: true,
        data: req.encode(),
      });
    }
  }

  publish(reg) {
    this.invoke(this._publishMethodName, [ reg ]);
    return this;
  }

  invoke(method, args, callback) {
    const oneway = !is.function(callback); // if no callback, means oneway
    const argLength = args.length;
    let data;
    // data:
    // +-----+---------------+-----+---------------+
    // | len |   arg1 body   | len |   arg2 body   |  ...
    // +-----+---------------+-----+---------------+
    if (argLength > 0) {
      let argsBufLength = 0;
      const arr = [];
      for (const arg of args) {
        const argBuf = this._transcode.encode(arg);
        const len = argBuf.length;
        const buf = Buffer.alloc(4 + len);
        buf.writeInt32BE(len, 0);
        argBuf.copy(buf, 4, 0, len);
        arr.push(buf);
        argsBufLength += (len + 4);
      }
      data = Buffer.concat(arr, argsBufLength);
    }
    const req = new Request({
      connObj: {
        type: 'invoke',
        method,
        argLength,
        oneway,
      },
      data,
      timeout: this.options.responseTimeout,
    });
    // send invoke request
    this.send({
      id: req.id,
      oneway,
      data: req.encode(),
    }, callback);
  }

  _registerChannel() {
    const req = new Request({
      connObj: {
        type: 'register_channel',
        channelName: this.options.name,
      },
      timeout: this.options.responseTimeout,
    });
    // send invoke request
    this.send({
      id: req.id,
      oneway: false,
      data: req.encode(),
    }, (err, data) => {
      if (err) {
        // if socket alive, do retry
        if (this._socket) {
          err.message = `register to channel: ${this.options.name} failed, will retry after 3s, ${err.message}`;
          this.logger.warn(err);
          // if exception, retry after 3s
          setTimeout(() => this._registerChannel(), 3000);
        } else {
          this.ready(err);
        }
        return;
      }
      const res = this._transcode.decode(data);
      if (res.success) {
        debug('[Follower:%s] register to channel: %s success', this.options.name, this.options.name);
        this.ready(true);
      } else {
        const error = new Error(res.error.message);
        Object.assign(error, res.error);
        this.ready(error);
      }
    });
  }

  _findMethodName(type) {
    for (const method of this.options.descriptors.keys()) {
      const descriptor = this.options.descriptors.get(method);
      if (descriptor.type === 'delegate' && descriptor.to === type) {
        return method;
      }
    }
    return null;
  }

  _handleRequest(req) {
    debug('[Follower:%s] receive req: %j from leader', this.options.name, req);
    const connObj = req.connObj || {};
    if (connObj.type === 'subscribe_result') {
      const result = this._transcode.decode(req.data);
      this.emit(connObj.key, result);
      this._subData.set(connObj.key, result);
      // feedback
      const res = new Response({
        id: req.id,
        timeout: req.timeout,
        connObj: { type: 'subscribe_result_res' },
      });
      this.send({
        id: req.id,
        oneway: true,
        data: res.encode(),
      });
    }
  }

  _connect(done) {
    if (!done) {
      done = err => {
        if (err) {
          this.ready(err);
        } else {
          // register to proper channel, difference type of client into difference channel
          this._registerChannel();
        }
      };
    }
    super._connect(done);
  }
}

module.exports = Follower;
